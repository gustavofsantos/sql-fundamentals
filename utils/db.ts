import { Client } from "pg";
import * as fs from "fs";

export async function createClient() {
  const client = new Client();
  await client.connect();

  return client;
}

export async function q(query: string) {
  const client = await createClient();
  const result = await client.query(query);
  await client.end();

  return result;
}

export async function createUsersTable() {
  const queryString = fs
    .readFileSync("./queries/create-users-table.sql", { encoding: "utf-8" })
    .toString();

  await q(queryString);
}

export async function createConstrainedUsersTable() {
  const queryString = fs
    .readFileSync("./queries/create-constrained-users-table.sql", {
      encoding: "utf-8"
    })
    .toString();

  await q(queryString);
}

export async function createConstrainedPurchasesTable() {
  const queryString = fs
    .readFileSync("./queries/create-constrained-purchases-table.sql", {
      encoding: "utf-8"
    })
    .toString();

  await q(queryString);
}

export async function dropUsersTable() {
  const queryString = fs
    .readFileSync("./queries/delete-users-table.sql", { encoding: "utf-8" })
    .toString();

  await q(queryString);
}

export async function dropPurchasesTable() {
  const queryString = fs
    .readFileSync("./queries/delete-purchases-table.sql", { encoding: "utf-8" })
    .toString();

  await q(queryString);
}

export async function selectAllUsers() {
  const queryString = fs
    .readFileSync("./queries/select-all-users.sql", { encoding: "utf-8" })
    .toString();

  return q(queryString);
}

export async function configureUuidExtension() {
  await q(`create extension "uuid-ossp";`);
}
