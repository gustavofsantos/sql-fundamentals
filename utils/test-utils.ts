import * as fs from "fs";
import { q } from "./db";

export async function insertUserTylerClark() {
  const query = fs
    .readFileSync("./queries/insert-user-tyler-clark.sql", {
      encoding: "utf-8"
    })
    .toString();

  await q(query);
}

export async function insertUserDebbieJones() {
  const query = fs
    .readFileSync("./queries/insert-user-debbie-jones.sql", {
      encoding: "utf-8"
    })
    .toString();

  await q(query);
}

export async function insertUsers() {
  await insertUserTylerClark();
  await insertUserDebbieJones();
}
