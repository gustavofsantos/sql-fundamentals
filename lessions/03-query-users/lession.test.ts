import { q, createUsersTable } from "../../utils/db";
import {
  insertUserDebbieJones,
  insertUserTylerClark
} from "../../utils/test-utils";

describe("Query users from Users table", () => {
  beforeEach(async () => {
    await createUsersTable();
    await insertUserTylerClark();
    await insertUserDebbieJones();
  });

  it("Should query all columns from Users table", async () => {
    const QUERY = `select * from Users;`;

    const result = await q(QUERY);
    expect(result.rows).toMatchSnapshot();
  });

  it("Should query first_name and last_name columns from Users table", async () => {
    const QUERY = `
      SELECT
        first_name, last_name
      FROM Users`;

    const result = await q(QUERY);
    expect(result.rows).toMatchSnapshot();
  });

  it("Should query first_name and last_name and give them FirstName and LastName alias", async () => {
    const QUERY = `select first_name as firstname, last_name as lastname from Users;`;

    const result = await q(QUERY);
    expect(result.rows).toMatchSnapshot();
  });

  it("Should count all rows in the Users table", async () => {
    const QUERY = `select count(*) from Users;`;

    const result = await q(QUERY);
    expect(result.rows).toMatchSnapshot();
  });

  it("Should get all distinct users by the first_name column", async () => {
    await q(
      `insert into Users (first_name, last_name) values ('tyler', 'clark');`
    );
    const QUERY = `select distinct(first_name) from Users`;

    const result = await q(QUERY);
    expect(result.rows).toMatchSnapshot();
  });

  it("Should count all distinct users by the last_name column", async () => {
    await q(
      `insert into Users (first_name, last_name) values ('tyler', 'clark');`
    );
    const QUERY = `select count(distinct(last_name)) from Users`;

    const result = await q(QUERY);
    expect(result.rows).toMatchSnapshot();
  });
});
