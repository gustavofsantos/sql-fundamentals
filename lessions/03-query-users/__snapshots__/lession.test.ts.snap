// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insert data into Users table Should count all distinct users by the last_name column 1`] = `
Array [
  Object {
    "count": "2",
  },
]
`;

exports[`Insert data into Users table Should count all rows in the Users table 1`] = `
Array [
  Object {
    "count": "2",
  },
]
`;

exports[`Insert data into Users table Should get all distinct users by the first_name column 1`] = `
Array [
  Object {
    "first_name": "tyler",
  },
  Object {
    "first_name": "debbie",
  },
]
`;

exports[`Insert data into Users table Should query all columns from Users table 1`] = `
Array [
  Object {
    "create_date": 2018-06-06T03:00:00.000Z,
    "first_name": "tyler",
    "last_name": "clark",
    "user_handle": "a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11",
  },
  Object {
    "create_date": 2018-06-06T03:00:00.000Z,
    "first_name": "debbie",
    "last_name": "jones",
    "user_handle": "a0eebc99-9c0b-42f8-bb6d-6bb9bd380a12",
  },
]
`;

exports[`Insert data into Users table Should query first_name and last_name and give them FirstName and LastName alias 1`] = `
Array [
  Object {
    "firstname": "tyler",
    "lastname": "clark",
  },
  Object {
    "firstname": "debbie",
    "lastname": "jones",
  },
]
`;

exports[`Insert data into Users table Should query first_name and last_name columns from Users table 1`] = `
Array [
  Object {
    "first_name": "tyler",
    "last_name": "clark",
  },
  Object {
    "first_name": "debbie",
    "last_name": "jones",
  },
]
`;

exports[`Query users from Users table Should count all distinct users by the last_name column 1`] = `
Array [
  Object {
    "count": "2",
  },
]
`;

exports[`Query users from Users table Should count all rows in the Users table 1`] = `
Array [
  Object {
    "count": "2",
  },
]
`;

exports[`Query users from Users table Should get all distinct users by the first_name column 1`] = `
Array [
  Object {
    "first_name": "tyler",
  },
  Object {
    "first_name": "debbie",
  },
]
`;

exports[`Query users from Users table Should query all columns from Users table 1`] = `
Array [
  Object {
    "create_date": 2018-06-06T03:00:00.000Z,
    "first_name": "tyler",
    "last_name": "clark",
    "user_handle": "a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11",
  },
  Object {
    "create_date": 2018-06-06T03:00:00.000Z,
    "first_name": "debbie",
    "last_name": "jones",
    "user_handle": "a0eebc99-9c0b-42f8-bb6d-6bb9bd380a12",
  },
]
`;

exports[`Query users from Users table Should query first_name and last_name and give them FirstName and LastName alias 1`] = `
Array [
  Object {
    "firstname": "tyler",
    "lastname": "clark",
  },
  Object {
    "firstname": "debbie",
    "lastname": "jones",
  },
]
`;

exports[`Query users from Users table Should query first_name and last_name columns from Users table 1`] = `
Array [
  Object {
    "first_name": "tyler",
    "last_name": "clark",
  },
  Object {
    "first_name": "debbie",
    "last_name": "jones",
  },
]
`;
